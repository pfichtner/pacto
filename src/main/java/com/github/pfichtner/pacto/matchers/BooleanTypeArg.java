package com.github.pfichtner.pacto.matchers;

import static java.lang.String.format;

public class BooleanTypeArg extends PactoMatcher<Boolean> {

	public BooleanTypeArg(boolean value) {
		super(value);
	}

	@Override
	public String toString() {
		return format("booleantype(%s)", getValue());
	}

}
