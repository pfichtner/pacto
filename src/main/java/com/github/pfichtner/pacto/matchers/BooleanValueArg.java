package com.github.pfichtner.pacto.matchers;

import static java.lang.String.format;

public class BooleanValueArg extends PactoMatcher<Boolean> {

	public BooleanValueArg(boolean value) {
		super(value);
	}

	@Override
	public String toString() {
		return format("booleanvalue(%s)", getValue());
	}

}
