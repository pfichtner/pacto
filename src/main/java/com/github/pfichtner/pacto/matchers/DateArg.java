package com.github.pfichtner.pacto.matchers;

import static java.lang.String.format;

import java.text.SimpleDateFormat;
import java.util.Date;

public class DateArg extends PactoMatcher<String> {

	private final Date example;

	public DateArg(String format, Date example) {
		super(format);
		this.example = example;
	}

	public Date example() {
		return example;
	}

	@Override
	public String toString() {
		var formatter = new SimpleDateFormat(value());
		return format("date(%s,%s)", value(), formatter.format(example));
	}

}
