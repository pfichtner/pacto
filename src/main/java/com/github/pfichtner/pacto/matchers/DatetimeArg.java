package com.github.pfichtner.pacto.matchers;

import static java.lang.String.format;

import java.text.SimpleDateFormat;
import java.util.Date;

public class DatetimeArg extends PactoMatcher<String> {

	private final Date example;

	public DatetimeArg(String format, Date example) {
		super(format);
		this.example = example;
	}

	public Date example() {
		return example;
	}

	@Override
	public String toString() {
		var formatter = new SimpleDateFormat(value());
		return format("datetime(%s,%s)", value(), formatter.format(example));
	}

}
