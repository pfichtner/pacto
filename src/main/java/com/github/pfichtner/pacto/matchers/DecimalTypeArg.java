package com.github.pfichtner.pacto.matchers;

public class DecimalTypeArg extends PactoMatcher<Double> {

	public DecimalTypeArg(double in) {
		super(in);
		withToStringFormat("decimal(%f)");
	}

}
