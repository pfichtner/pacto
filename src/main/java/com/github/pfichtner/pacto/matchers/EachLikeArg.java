package com.github.pfichtner.pacto.matchers;

public class EachLikeArg extends PactoMatcher<Object> {

	private final Object value;
	private Integer max;
	private Integer min;

	public EachLikeArg(Object value) {
		this.value = value;
	}

	public Object getValue() {
		return value;
	}

	public EachLikeArg max(int max) {
		this.max = max;
		return this;
	}

	public Integer getMax() {
		return max;
	}

	public EachLikeArg min(int min) {
		this.min = min;
		return this;
	}

	public Integer getMin() {
		return min;
	}

	@Override
	public String toString() {
		if (max != null) {
			return String.format("maxArrayLike(%s, %d)", value, max);
		}
		if (min != null) {
			return String.format("minArrayLike(%s, %d)", value, min);
		}
		return String.format("eachLike(%s)", value);
	}

}
