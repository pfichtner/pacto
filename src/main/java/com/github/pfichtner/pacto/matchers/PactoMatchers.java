package com.github.pfichtner.pacto.matchers;

import static org.mockito.internal.progress.ThreadSafeMockingProgress.mockingProgress;

import java.lang.reflect.Array;
import java.util.List;
import java.util.Set;

import org.mockito.ArgumentMatcher;

public final class PactoMatchers {

	public static final boolean DEFAULT_BOOLEAN_VALUE = true;
	public static final String DEFAULT_STRING_VALUE = "string";
	public static final int DEFAULT_INTEGER_VALUE = 42;
	public static final float DEFAULT_DECIMAL_VALUE = 12.345F;
	public static final String DEFAULT_HEX_VALUE = "1234a";

	private PactoMatchers() {
		super();
	}

	public static boolean booleanType() {
		return booleanType(DEFAULT_BOOLEAN_VALUE);
	}

	public static boolean booleanType(boolean in) {
		reportMatcher(new BooleanTypeArg(in));
		return in;
	}

	public static boolean booleanValue(boolean in) {
		reportMatcher(new BooleanValueArg(in));
		return in;
	}

	public static String stringType() {
		return stringType(DEFAULT_STRING_VALUE);
	}

	public static String stringType(String in) {
		reportMatcher(new StringTypeArg(in));
		return in;
	}

	public static String stringMatcher(String regex, String in) {
		reportMatcher(new StringMatcherArg(regex, in));
		return in;
	}

	public static int integerType() {
		return integerType(DEFAULT_INTEGER_VALUE);
	}

	public static int integerType(int in) {
		reportMatcher(new IntegerTypeArg(in));
		return in;
	}

	public static long integerType(long in) {
		reportMatcher(new IntegerTypeArg(in));
		return in;
	}

	public static float decimalType() {
		return decimalType(DEFAULT_DECIMAL_VALUE);
	}

	public static double decimalType(double in) {
		reportMatcher(new DecimalTypeArg(in));
		return in;
	}

	public static float decimalType(float in) {
		reportMatcher(new DecimalTypeArg(in));
		return in;
	}

	public static <T extends Number> T numberType(T in) {
		reportMatcher(new NumberTypeArg(in));
		return in;
	}

	public static String hex() {
		return hex(DEFAULT_HEX_VALUE);
	}

	public static String hex(String in) {
		reportMatcher(new HexArg(in));
		return in;
	}

	public static <T> T nullValue() {
		reportMatcher(new NullValueArg());
		return null;
	}

	public static <T> T[] maxArrayLike(T value, int max) {
		return eachLike(value, new EachLikeArg(value).max(max));
	}

	public static <T> T[] minArrayLike(T value, int min) {
		return eachLike(value, new EachLikeArg(value).min(min));
	}

	public static <T> T[] eachLike(T value) {
		return eachLike(value, new EachLikeArg(value));
	}

	@SuppressWarnings("unchecked")
	private static <T> T[] eachLike(T value, EachLikeArg matcher) {
		reportMatcher(matcher);
		T[] values = (T[]) Array.newInstance(value.getClass(), 1);
		values[0] = value;
		return values;
	}

	public static class Lists {
		public static <T> List<T> maxArrayLike(T value, int max) {
			return eachLike(value, new EachLikeArg(value).max(max));
		}

		public static <T> List<T> minArrayLike(T value, int min) {
			return eachLike(value, new EachLikeArg(value).min(min));
		}

		public static <T> List<T> eachLike(T value) {
			return eachLike(value, new EachLikeArg(value));
		}

		private static <T> List<T> eachLike(T value, EachLikeArg matcher) {
			reportMatcher(matcher);
			return List.of(value);
		}
	}

	public static class Sets {
		public static <T> Set<T> maxArrayLike(T value, int max) {
			return eachLike(value, new EachLikeArg(value).max(max));
		}

		public static <T> Set<T> minArrayLike(T value, int min) {
			return eachLike(value, new EachLikeArg(value).min(min));
		}

		public static <T> Set<T> eachLike(T value) {
			return eachLike(value, new EachLikeArg(value));
		}

		private static <T> Set<T> eachLike(T value, EachLikeArg matcher) {
			reportMatcher(matcher);
			return Set.of(value);
		}
	}

	private static void reportMatcher(ArgumentMatcher<?> matcher) {
		mockingProgress().getArgumentMatcherStorage().reportMatcher(matcher);
	}

}
