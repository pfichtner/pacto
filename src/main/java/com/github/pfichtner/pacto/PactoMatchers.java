package com.github.pfichtner.pacto;

import static org.mockito.internal.progress.ThreadSafeMockingProgress.mockingProgress;

import org.mockito.ArgumentMatcher;

import com.github.pfichtner.pacto.matchers.IntegerTypeArg;
import com.github.pfichtner.pacto.matchers.RegexArg;
import com.github.pfichtner.pacto.matchers.StringTypeArg;

public class PactoMatchers {

	public static final String DEFAULT_STRING_VALUE = "string";
	public static final int DEFAULT_INTEGER_VALUE = 42;

	public static String stringType() {
		return stringType(DEFAULT_STRING_VALUE);
	}

	public static String stringType(String in) {
		StringTypeArg matcher = new StringTypeArg(in);
		reportMatcher(matcher);
		return in;
	}

	public static String regex(String regex, String in) {
		RegexArg matcher = new RegexArg(regex, in);
		reportMatcher(matcher);
		return in;
	}

	public static int integerType() {
		return integerType(DEFAULT_INTEGER_VALUE);
	}

	public static int integerType(int in) {
		IntegerTypeArg matcher = new IntegerTypeArg(in);
		reportMatcher(matcher);
		return in;
	}

	private static void reportMatcher(ArgumentMatcher<?> matcher) {
		mockingProgress().getArgumentMatcherStorage().reportMatcher(matcher);
	}

}
