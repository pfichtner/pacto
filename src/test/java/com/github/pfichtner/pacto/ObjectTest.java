package com.github.pfichtner.pacto;

import static com.github.pfichtner.pacto.Pacto.spec;
import static com.github.pfichtner.pacto.PactoDslBuilder.dslFrom;
import static com.google.gson.JsonParser.parseString;
import static org.approvaltests.JsonApprovals.verifyAsJson;

import java.lang.reflect.Array;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Date;
import java.util.List;
import java.util.Set;
import java.util.TreeMap;
import java.util.UUID;

import org.junit.jupiter.api.Test;

import com.github.pfichtner.pacto.StaticMethodInvoker.ArgumentProvider;
import com.github.pfichtner.pacto.matchers.PactoMatchers;
import com.github.pfichtner.pacto.matchers.TestTarget;
import com.github.pfichtner.pacto.testdata.Foo;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;

import au.com.dius.pact.consumer.dsl.DslPart;

class ObjectTest {

	@Test
	void verifyObject() throws Exception {
		TestTarget spec = spec(new TestTarget());
		new StaticMethodInvoker(PactoMatchers.class, spec, argumentProvider()).invoke();
		DslPart dslFrom = dslFrom(spec);

		Gson gson = new GsonBuilder().setPrettyPrinting().create();
		var root = new JsonObject();
		root.add("body", parseString(dslFrom.toString()).getAsJsonObject());
		root.add("matchingRules", gson.toJsonTree(new TreeMap<>(dslFrom.getMatchers().getMatchingRules())));
		verifyAsJson(root);
	}

	private static ArgumentProvider argumentProvider() {
		return (m, t) -> {
			if (t == String.class) {
				if ("hex".equals(m)) {
					return "00FF";
				} else if ("uuid".equals(m)) {
					return "e2490de5-5bd3-43d5-b7c4-526e33f71304";
				} else if ("ipAddress".equals(m)) {
					return "127.0.0.1";
				} else if ("date".equals(m)) {
					return "yyyy";
				} else if ("time".equals(m) || "datetime".equals(m)) {
					return "HH:MM";
				}
				return "example";
			} else if (t == int.class || t == Integer.class) {
				return 42;
			} else if (t == long.class || t == Long.class) {
				return 123L;
			} else if (t == double.class || t == Double.class) {
				return 3.14;
			} else if (t == float.class || t == Float.class) {
				return 2.71f;
			} else if (t == boolean.class || t == Boolean.class) {
				return true;
			} else if (t == UUID.class) {
				return UUID.randomUUID();
			} else if (t == Date.class) {
				return new Date();
			} else if (t == java.time.LocalDate.class) {
				return LocalDate.now();
			} else if (t == java.time.LocalDateTime.class) {
				return LocalDateTime.now();
			} else if (t.isArray()) {
				return Array.newInstance(t.getComponentType(), 0);
			} else if (List.class.isAssignableFrom(t)) {
				return List.of();
			} else if (Set.class.isAssignableFrom(t)) {
				return Set.of();
			} else {
				return spec(new Foo());
			}
		};
	}
}
