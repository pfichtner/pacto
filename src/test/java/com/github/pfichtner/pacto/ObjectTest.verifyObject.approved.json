{
  "body": {
    "arrayArg": [
      {}
    ],
    "dateArg": "13:08:03",
    "doubleArg": 0.34667498058347623,
    "doubleWrapperArg": 0.34667498058347623,
    "floatArg": 0.505189836025238,
    "floatWrapperArg": 0.50518984,
    "intArg": 1881612699,
    "integerWrapperArg": 1881612699,
    "localDateArg": "42590977-03-08",
    "localDateTimeArg": "08:46:16",
    "longArg": -1830391751231062889,
    "longWrapperArg": -1830391751231062889,
    "numberArg": -1830391751231062889,
    "objectArg": "2f6086d6-bf2c-38b4-a1e9-2b740fc93f95",
    "stringArg": "o9zWFfoviShcKsgfZT6rCJMkn5MPQH2y",
    "uuidArg": "e2490de5-5bd3-43d5-b7c4-526e33f71304"
  },
  "matchingRules": {
    ".arrayArg": {
      "rules": [
        {},
        {
          "max": -1819415847
        },
        {
          "min": 205323323
        },
        {
          "min": 765357112,
          "max": 765357112
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".dateArg": {
      "rules": [
        {
          "format": "yyyy-MM"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "hh:mm"
        },
        {
          "format": "yyyy-MM-dd HH:mm:ss"
        },
        {
          "format": "hh:mm"
        },
        {
          "format": "HH:mm:ss"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".doubleArg": {
      "rules": [
        {
          "numberType": "DECIMAL"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".doubleWrapperArg": {
      "rules": [
        {
          "numberType": "DECIMAL"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".floatArg": {
      "rules": [
        {
          "numberType": "DECIMAL"
        },
        {
          "numberType": "DECIMAL"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".floatWrapperArg": {
      "rules": [
        {
          "numberType": "DECIMAL"
        },
        {
          "numberType": "DECIMAL"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".intArg": {
      "rules": [
        {},
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".integerWrapperArg": {
      "rules": [
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".localDateArg": {
      "rules": [
        {
          "format": "yyyy-MM"
        },
        {
          "format": "yyyy-MM-dd"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".localDateTimeArg": {
      "rules": [
        {
          "format": "hh:mm"
        },
        {
          "format": "HH:mm:ss"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".longArg": {
      "rules": [
        {},
        {
          "numberType": "INTEGER"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".longWrapperArg": {
      "rules": [
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".numberArg": {
      "rules": [
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        },
        {
          "numberType": "NUMBER"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".objectArg": {
      "rules": [
        {},
        {},
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "numberType": "DECIMAL"
        },
        {
          "numberType": "DECIMAL"
        },
        {
          "numberType": "DECIMAL"
        },
        {},
        {},
        {
          "regex": "[0-9a-fA-F]+"
        },
        {
          "regex": "[0-9a-fA-F]+"
        },
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        },
        {
          "value": "zwQB5Kp2mLDs26ahJnyJkJ5OfC7KY5Aa"
        },
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        },
        {
          "numberType": "INTEGER"
        },
        {
          "regex": "(\\d{1,3}\\.)+\\d{1,3}"
        },
        {
          "regex": ".*\\/()$"
        },
        {},
        {},
        {},
        {},
        {},
        {
          "format": "\u0027T\u0027HH:mm:ss"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "format": "\u0027T\u0027HH:mm:ss"
        },
        {
          "format": "yyyy-MM-dd"
        },
        {
          "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
        },
        {
          "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
        },
        {
          "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    },
    ".uuidArg": {
      "rules": [
        {
          "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
        },
        {
          "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
        },
        {
          "regex": "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}"
        }
      ],
      "ruleLogic": "AND",
      "cascaded": false
    }
  }
}
